### ISSUE TITLE
Founding fathers logic gives disadvantage to human player

### ISSUE THREAD

**Issue by jo-oe** [We-the-People-civ4col-mod/Mod#770]
I was trying to understand why the (human) player still has a disadvantage compared to AI when choosing founding fathers (FFs). It seems that the player only gets a chance to bid for an FF if no AI has the necessary FF points available.
This is because of the following logic:
- FF recruitment is done during team turns, which happen before player turns
- TEAM TURN -> CvTeam::testFoundingFather(): for AI, an FF that is recruitable (canConvinceFather == true) will immediately be recruited
- TEAM TURN -> CvTeam::testFoundingFather(): for a team with human players, the player will get a popup to decide on recruitment during the next player turn
--- all teams have now taken their turn
- PLAYER TURN -> CvDLLButtonPopup::OnFocus(): before showing the popup where the player can decide on recruiting, there is another check of canConvinceFather() to cancel the popup. As AI was probably able to recruit the founding father by now (because this happened during team turns), it is not available anymore.

This jeopardizes my recent attempt to give the player a chance to bid for FF freed from eliminated AI players ...

I had the idea to move player turns inside the team turns like (pseudocode):
```
CvGame::doTurn()
{
...
foreach(eTeam)
 {
 eTeam.doTurn() 
 }
...
}

CvTeam::doTurn()
{
...
testFoundingFather();
foreach(ePlayer)
 {
 if (ePlayer.Team() == this)
   {
   ePlayer.doTurn();
   }
  }
...
}
```

What do you think of this approach? I actually think it would be more logical this way, but it would probably mean a significant change to the general sequence of things happening during game turns.
I have no idea though if this actually works and what consequences it might have ...

An alternative would be to do Team turns inside Player turns (e.g. when the first player in a team takes their turn), which would be against my logical understanding though.

**Comment by jo-oe** [comment]
:
CvGame::doTurn():
1. loops through all teams and calls CvTeam::doTurn for each
2. loops through either
 a) - if MPOPTION_SIMULTANEOUS_TURNS - all players in randomized order and calls CvPlayer::setTurnActive(true) or
 b) - if isSimultaneousTeamTurns() - all teams and calls CvTeam::setTurnActive(true) or
 c) - if neither a) nor b) - all players and calls CvPlayer::setTurnActive(true)

CvTeam::setTurnActive()
 loops through all players and calls CvPlayer::setTurnActive(true) for each

CvPlayer::setTurnActive()
 - calls CvPlayer::doTurn(), but in different locations depending on MPOPTION_SIMULTANEOUS_TURNS
 - if isSimultaneousTeamTurns() - loops through all teams and calls CvTeam::setTurnActive() for each

HUH???

**Comment by Nightinggale** [comment]
We could also change the approach to how FF joins teams. We can make teams send a request to an FF to join. When all players have clicked end turn, the game goes through the list and make a list of interested teams, which can get the FF. It then picks a random one. This way it's fair in the sense that the order of players/teams will not influence the chance to get an FF.

The downside of this is a human player is asked and then the FF will join the following turn.

**Comment by jo-oe** [comment]
> We could also change the approach to how FF joins teams. We can make teams send a request to an FF to join. When all players have clicked end turn, the game goes through the list and make a list of interested teams, which can get the FF. It then picks a random one. This way it's fair in the sense that the order of players/teams will not influence the chance to get an FF.
> 
> The downside of this is a human player is asked and then the FF will join the following turn.

Sounds like an interesting option to me, but would probably require some UI tweaking I don't yet feel comfortable about (Python is still on my to-learn list ;)
Also there is another problem your suggestion might cause: This logic would only allow for one FF to be assigned per turn. In some games (especially non-marathon), total FF point gain of all players might be higher than one FF per turn, so we would delay the spending of all points by a significant amount of turns.
(because we would probably run into complicated issues if we allow several FFs to be requested per turn and then one is randomly assigned to the player so he does not have all necessary FF points left to buy another FF requested. Then we have to re-roll the random number etc. making the whole logic convoluted and messy)

**Comment by jo-oe** [comment]
I'd also like to leave the link devolution posted in discord here for reference:
https://github.com/caveman2cosmos/Caveman2Cosmos/wiki/Anatomy-of-a-Turn

> "The code in question seem to be the definition of messy code."
Exactly! Sounds like we should fix it ;)

**Comment by jo-oe** [comment]
Remark: This is not exactly a regression compared to the behaviour before I changed the FF logic.
Before the changes:
AIs would get all FFs they could afford within a single turn (giving an advantage to AIs depending on player ordering)
If there were FFs remaining which could not be afforded by any AI but by the human player, the human player would now be able to bid on them.

After my changes:
The AIs each get one FF per turn until they run of FF points.
In the first turn when no AI has FF points to get a new FF, the player has a chance to bid on one of the remaining ones.

Under the condition that all AIs eventually run out of FF points, the only difference now is that it takes a few turns for the AIs to select all their FFs, when before they would get them all in one turn.

**The problem is that a FF is chosen to be offered the the player's team which is then taken by an AI team (during the team turn) before the player can choose to recruit the FF (during the player turn).**

**Comment by raystuttgart** [comment]
Nothing I read here sounds like a blocking bug.
But I had understood that you are still trying to improve the code.

**Comment by raystuttgart** [comment]
**By the way:**

The suggestion of @Nightinggale does not sound that bad.
Why not have something like a FF bidding?

Have all players / teams cast their bet during the team logic.
And then during the game logic select the team that has most FF-points that fit to the FF.

That way it would be fair, since the tean with the most FF-points would win and not the one which comes first in turn order.
For the unlikely case where 2 teams have the same amount of FF-Points then turn order may determine who gets it.

**Comment by Nightinggale** [comment]
> It may require some caching of data, but generally it should be doable.
(The "bets" for FFs would need to be cached so the game turn logic can read it.

That's just a vector of <FatherTypes, TeamTypes>. We can even sort it easily meaning it will be fine adding them out of order.

**Comment by jo-oe** [comment]
That would be a proper fix to the problem. It would also be nicer than just being offered a FF and not knowing the alternatives.
There is a few things we should think about though:
1) it might get really complicated if you make several bids which surpass the amount of total FF points. Now you succeed with one of your bids and then lack the FF points for another bid, we need to check all the other bids if they are still valid etc.
2) I would like to have some sort of notification that you have the necessary FF points ready to bid. Otherwise always checking for them will be quite annoying.

**Comment by raystuttgart** [comment]
Actually the "bidding" should simply still be triggered automatically by the FF pop-up just like now.
So no new notification or anything like that needs to happen.

In the end we will simply have a list of all FFs that have been bid on by all players.
And then during the game turn it will be looped and the FF will always go to the player that has the highest point score matching the FF.

When you have bid on an FF the UI should display a text message that you asked the FF to join your cause.
Also if you have **not** been successful with a bid, you get a message that the FF has chosen to join another player.

----

**In other words:**
UI will still be the same.
Just the logic in the background will be different.

**Comment by Nightinggale** [comment]
1: that seems fairly easy. Once the bidding should be resolved, just sort the list by FF and in case of a tie, by TeamTypes. Next kill duplicates (easy now that it's sorted as that's just kill those, which are identical to the previous one). Now go through the list and for each FF mentioned, take all the teams bidding for it, check that the offer is still valid and then for valid ones add them to a list. Then pick a random one from that list and clear it, making it ready for the next FF.

This approach relies heavily on standard C++ functions (sort etc) and calls to out code (can have FF). This means the code to resolve the bidding should be possible to accomplish fairly easy.

**Comment by raystuttgart** [comment]
Yeah, it really does not seem to be that complicated actually.
And it even sounds more fun and fair than the current system.

**Comment by jo-oe** [comment]
I'm not completely convinced that the actual player experience will be intuitive if we keep the delay between the choice in the UI and the actual recruitment. Especially in a case when there are more than 1 FF available to bid on.

**Comment by raystuttgart** [comment]
The player experience should be intuitive if you also post a text message in the screen that explains it.

**Comment by jo-oe** [comment]
The situation that worries me is when several players have the FF points available to bid on multiple FFs.
E.g.
Player1:
pointsA: 1000
pointsB: 1000
pointsC: 1000

Player2:
pointsA: 1000
pointsB: 1000
pointsC: 1000

Now you have FF worth 100, 200, 300, 400 points of either A+B or A+C

Each player can either bid or pass on 8 different FFs. But you are not shown the other choices you will have during this turn if you either pass or not. So to the player it will probably feel like a black box.

The messages would only appear on the next turn, probably obscured by the 10s of other messages during the AI turn.

**Comment by jo-oe** [comment]
The problem is that the **actual recruitment will only occur during the next AI turn**, with a lot of stuff happening on each turn and a lot of sounds played due to all the battles etc.

**In the end it boils down to: I don't feel like I will be able to implement this in a way I would like myself as a player.**
That doesn't mean I want to keep anyone else from doing it of course.

**Comment by Nightinggale** [comment]
When bidding is resolved, we can create a popup for involved teams/players telling the outcome for each FF they bid on. We can also change the text in the popup asking if you want an FF to indicate that it's an attempt to get them to join next turn.

**Comment by raystuttgart** [comment]
And of course once you have acquired an FF your FF points will be reduced.
That also means the next bid you have placed is less likely to succeed.
Thus the more bids you have the less likely it is that they will all be successfull.

**Comment by raystuttgart** [comment]
Also it is exrtremely unlikely that 2 players both have exactly 1000 FF points.

e.g. 

-----

Step 1:

Both players bid on FF X for 500 FF points.
Both players bid on FF Y for 200 FF points

Player A will e.g. have 1005 FF points.
Player B will e.g. have 1021 FF points

-----

Step 2:

Player B wins FF X since he has more FF points. (1021 > 1005)

Player A still has 1005 FF points
Now Player 2 only has 521 FF points

-----

Step 3:

Player A wins FF Y since he has more FF ponts. (1005 > 521)

Player A has 805 FF points left
Player B has 521 points left

--> Player A would most likely also win the 3rd bet if there was any.
(805 > 521)

**Comment by jo-oe** [comment]
Sorry, but that's not my point, it is the same in your example.
My point is that it is very confusing if you are first shown several popups and don't oversee the consequences. Neither you know at this point which other FFs are available during the same turn nor if your FF points will be sufficient for them if you bid or not.
You have to make choices at the beginning of a turn and will see the result buried in the 100 messages at the beginning of the next turn.

**Comment by raystuttgart** [comment]
It is simply a bidding that you may win or lose.

However it is unlikely that you may bid on more than 1 FF a turn anyways.
(It will probably only happen a few times a game.)

-------

Of course it was just a suggestion.
Feel free to implement whatever you like. 👍

**Comment by raystuttgart** [comment]
**By the way:**
Even **concurrent biddings** of different players for the same FF should be **extremely rare**.
So it makes me think if such a bidding system is even worth the effort ...

**Comment by jo-oe** [comment]
That is exactly the point where I encountered this bug, see my original post:
> This jeopardizes my recent attempt to give the player a chance to bid for FF freed from eliminated AI players ...

**Comment by raystuttgart** [comment]
But why not still check the FF points first?

1) If FF points of AI for that father are higher than FF points of Human Player --> give it to AI
2) If FF points of AI for that father are lower than FF points of Human Player --> do NOT give it to AI and let Human have a shot

**Comment by jo-oe** [comment]
And 3. If FF points for AI are lower than FF points of Human Player, but Human Player has selected ignore --> give it to AI
Though we would need to make a list for each FF somewhere, testing which player has the highest FF score for this FF __and__ sufficient FF points to pay for it. That would probably have to happen during CvGame::doTurn. And in which class would we store that list? There is no class for FFs yet - might be time to implement one ...

**Comment by raystuttgart** [comment]
You do not need to make a cached list for that.
You can simply read the FF points in a loop logic.

**Comment by jo-oe** [comment]
But then we would have at least three nested loops
1. loop teams (doTurn)
2. loop FFs
3. loop teams again to check which team deserves this FF most

**Comment by raystuttgart** [comment]
I do not really get the problem of writing and calling a logic in CvGame like this:
**getTeamThatDeservesFFMost(eFFType)**

But I also have not yet read the code.
But something like that should be possible by looping.
